### PR TITLE
gpu support

### PR DESCRIPTION
**Pull request by Qfl3x**

### CODE REVIEW

File: src/data/prepare_data.jl [hunk 1]
```diff
@@ -1,9 +1,12 @@
...
 
-function prepare_data(hm, data::KeyedArray; kwargs...)
-    predictors_forcing, targets = get_prediction_target_names(hm)
+function prepare_data(hm, data::KeyedArray; cfg=DataConfig(), kwargs...)
+    predictors, forcings, targets = get_prediction_target_names(hm)
     # KeyedArray: use () syntax for views that are differentiable
-    return (data(predictors_forcing), data(targets))
+    dev = cfg.gdev
+    targets_nt = NamedTuple([target => dev(Array(data(target))) for target in targets])
+    forcings_nt = NamedTuple([forcing => dev(Array(data(forcing))) for forcing in forcings])
```
Comment (lazarusA) [inline src/data/prepare_data.jl]: we should do `dev`/ `Array` at the batch loader level. Up to this point data could still be lazy.

File: src/training/early_stopping.jl [hunk 2]
```diff
@@ -70,16 +70,16 @@ function best_or_final(stopper::EarlyStopping, ps, st, cfg::TrainConfig)
...
 
-function build_results(model, history::TrainingHistory, stopper::EarlyStopping, ps, st, x_train, y_train, x_val, y_val)
+function build_results(model, history::TrainingHistory, stopper::EarlyStopping, ps, st, x_train, forcings_train, y_train, x_val, forcings_val, y_val, cfg::TrainConfig)
     target_names = model.targets
 
     # final predictions in test mode
-    ŷ_train, _ = model(x_train, ps, LuxCore.testmode(st))
-    ŷ_val, _ = model(x_val, ps, LuxCore.testmode(st))
+    ŷ_train, _ = model((cfg.cdev(x_train), cfg.cdev(forcings_train)), cfg.cdev(ps), LuxCore.testmode(st))
+    ŷ_val, _ = model((cfg.cdev(x_val), cfg.cdev(forcings_val)), cfg.cdev(ps), LuxCore.testmode(st))
```
Comment (lazarusA) [inline src/training/early_stopping.jl:78]: I think we can evaluate this still on the GPU side and just pipe the result of into the `cfg.dev` function.

File: src/training/epoch.jl [hunk 3]
```diff
@@ -1,23 +1,26 @@
-function run_epoch!(loader, model, ps, st, opt_state, cfg::TrainConfig)
+function run_epoch!(loader, model, ps, st, train_state, cfg::TrainConfig)
     loss_fn = build_loss_fn(model, cfg)
 
-    for (x, y) in loader
-        is_no_nan = valid_mask(y)
-        isnothing(is_no_nan) && continue
+    for (x, y) in cfg.gdev(loader)
+        is_no_nan = falses(length(first(y))) |> cfg.gdev
```
Comment (lazarusA) [inline src/training/epoch.jl]: the `cfg.gdev(loader)` is already moving the data into the gpu device, the second line should be an operation on the gpu side already, hence `|> cfg.gdev` should not be needed, in principle

File: src/training/initialization.jl [hunk 4]
```diff
@@ -16,15 +16,16 @@ end
 
 function init_model_state(model, cfg::TrainConfig)
     if isnothing(cfg.train_from)
-        ps, st = LuxCore.setup(Random.default_rng(), model)
+        ps, st = LuxCore.setup(Random.default_rng(), model) |> cfg.gdev
+        ps = ps |> cfg.gdev
```
[the remark this answers is not in the document]
Comment (lazarusA) [reply src/training/initialization.jl]: indeed, line 20 is not needed here.

**Comment by lazarusA** [comment]
comments are being addressed in #257

**Comment by Qfl3x** [comment]
Main things I changed here are:

1.  Masking is now pre-computed. Found that it was a bit expensive previously.
2. cleaner device switching, 
3. Changing previous tests to conform to the new interface

I think the main thing that needs work on here and/or #257 is device switching, namely that device switch should happen at the batch level. CUDA needs contiguous arrays anyways to work, so views are a no no. Meaning we have to allocate a new array at every batch anyways, so may as well do the device switching there.

I worked here before I saw all the comments. So I'll now switch to #257

**Comment by lazarusA** [comment]
We can also continue here, whatever merge branches is easier. Both are still open. I just didn't wanna to have merge conflicts just in case you did local work 😌, as you have done 👍.

File: test/test_compute_loss.jl [hunk 5]
```diff
@@ -8,9 +8,9 @@ using DataFrames
 
 @testset "_compute_loss" begin
     # Test data setup
-    ŷ = Dict(:var1 => [1.0, 2.0, 3.0], :var2 => [2.0, 3.0, 4.0])
-    y(target) = target == :var1 ? [1.1, 1.9, 3.2] : [1.8, 3.1, 3.9]
```
Comment (lazarusA) [inline test/test_compute_loss.jl:12]: doing `y(target)` was a intended to mirror AxisKeys syntax, although admittedly it would be better to do an independent test for that, and similarly for DD.

File: test/test_compute_loss.jl [hunk 6]
```diff
@@ -108,10 +88,30 @@ using DataFrames
...
+    #     # Test single predefined loss
+    #     loss = _compute_loss(ŷ_dim, y_dim, y_nan_dim, targets, :mse, sum)
+    #     @test loss isa Number
+    #
+    #     # Test multiple predefined losses
+    #     losses = _compute_loss(ŷ_dim, y_dim, y_nan_dim, targets, [:mse, :mae], sum)
+    #     @test losses isa NamedTuple
+    #     @test haskey(losses, :mse)
+    #     @test haskey(losses, :mae)
+    # end
```
Comment (lazarusA) [inline test/test_compute_loss.jl]: before merging, we'll these back. The gotcha here is the hardcode `Dim{:variable}`.

File: src/data/prepare_data.jl [hunk 7]
```diff
@@ -1,9 +1,12 @@
...
+    predictors, forcings, targets = get_prediction_target_names(hm)
     # KeyedArray: use () syntax for views that are differentiable
-    return (data(predictors_forcing), data(targets))
+    X_arr = Array(data(predictors)) 
+    targets_nt = NamedTuple([target => Array(data(target)) for target in targets])
+    forcings_nt = NamedTuple([forcing => Array(data(forcing)) for forcing in forcings])
+    return ((X_arr, forcings_nt), targets_nt)
 end
 
 function prepare_data(hm, data::AbstractDimArray; kwargs...)
```
Comment (lazarusA) [inline src/data/prepare_data.jl:21]: we are missing the updates for this dispatch.

**Comment by lazarusA** [comment]
On regards to data:

- csv files - > DataFrames (non-AD) -> AxisKeys (AD)
- zarr / nc -> YAXArray (DD which now also supports AD), less tested, but examples in main with direct DD types are working.

File: src/data/split_data.jl [hunk 8]
```diff
@@ -50,9 +50,15 @@ function split_data(
...
-        x_val, y_val = view_end_dim(x_all, val_idx), view_end_dim(y_all, val_idx)
-        return (x_train, y_train), (x_val, y_val)
+        x_train, forcings_train, y_train = collect_end_dim(x_all, train_idx),collect_end_dim(forcings_all, train_idx), collect_end_dim(y_all, train_idx)
+        x_val, forcings_val, y_val = collect_end_dim(x_all, val_idx),collect_end_dim(forcings_all, val_idx), collect_end_dim(y_all, val_idx)
+        x_train = x_train |> cfg.gdev
+        forcings_train = forcings_train |> cfg.gdev
+        y_train = y_train |> cfg.gdev
+        x_val = x_val |> cfg.gdev
+        forcings_val = forcings_val |> cfg.gdev
+        y_val = y_val |> cfg.gdev
```
Comment (lazarusA) [inline src/data/split_data.jl]: maybe it would be best if all these pipes move out to the inner loop in the batches?

File: src/training/epoch.jl [hunk 9]
```diff
@@ -1,30 +1,28 @@
-function run_epoch!(loader, model, ps, st, opt_state, cfg::TrainConfig)
+function run_epoch!(loader, model, ps, st, train_state, cfg::TrainConfig)
     loss_fn = build_loss_fn(model, cfg)
 
     for (x, y) in loader
-        is_no_nan = valid_mask(y)
```
Comment (lazarusA) [inline src/training/epoch.jl:5]: this was skipping the case of empty batches (all samples were NaN). Which, it should not happen, but you never know what the user will provide. Maybe we should keep this, or, was this moved somewhere else?

File: test/test_compute_loss.jl (hunk 5 above)
Comment (Qfl3x) [reply test/test_compute_loss.jl:12]: For now I've written the interface for loss functions to accept named tuples. So I may have to add a dispatch for callables.

File: src/data/prepare_data.jl (hunk 7 above)
Comment (Qfl3x) [reply src/data/prepare_data.jl:21]: Dispatch for DimArray updated. Should work on training now, other than that it may crash if you use DimArray for anything else.

File: src/training/epoch.jl [hunk 10]
```diff
@@ -2,6 +2,10 @@ function run_epoch!(loader, model, ps, st, train_state, cfg::TrainConfig)
     loss_fn = build_loss_fn(model, cfg)
 
     for (x, y) in loader
```
Comment (lazarusA) [inline src/training/epoch.jl:15]: I think is better to do `cfg.gdev(loader)`, is just a wrapper at this point, I think. Then, when loop over, the  `(x,y)` will be sent to the gpu. If it works now, we can come back to this later.

File: src/training/epoch.jl (hunk 9 above)
Comment (Qfl3x) [reply src/training/epoch.jl:5]: Added it back. Though note that the mask works differently than previously. Now there is a mask per target.

Comment (lazarusA) [reply src/training/epoch.jl:5]: yes, mask per target is the right move forward! thanks.

File: src/data/prepare_data.jl (hunk 7 above)
Comment (Qfl3x) [reply src/data/prepare_data.jl:21]: I just confirmed, DimArray works in training (goes down to a named tuple with lazy DimArray values).

File: test/test_compute_loss.jl (hunk 6 above)
Comment (Qfl3x) [reply test/test_compute_loss.jl]: Since training works with named tuples now, isn't this redundant? If you want it added I could always dispatch on it.

Comment (lazarusA) [reply test/test_compute_loss.jl]: let's add a dispatch.

File: src/training/epoch.jl (hunk 10 above)
Comment (Qfl3x) [reply src/training/epoch.jl:15]: It works to do `cfg.gdev(loader)`  except for dimensional data, it's still lazy and is not a simple array like the rest.

File: test/test_compute_loss.jl (hunk 6 above)
Comment (Qfl3x) [reply test/test_compute_loss.jl]: Fixed it, the dispatch already existed.

File: test/test_compute_loss.jl (hunk 5 above)
Comment (Qfl3x) [reply test/test_compute_loss.jl:12]: These should work again.

**Comment by Qfl3x** [comment]
Most things have been addressed. Docs still have problems.

**Comment by lazarusA** [comment]
I suppose docs issues are related to the new prepare data step.

File: src/training/epoch.jl [hunk 11]
```diff
@@ -39,17 +46,15 @@ function build_loss_fn(model, cfg::TrainConfig)
     )
 end
 
-function evaluate_epoch(model, x_train, y_train, x_val, y_val, ps, st, init::EpochSnapshot, cfg::TrainConfig)
-    is_no_nan_t = .!isnan.(y_train)
-    is_no_nan_v = .!isnan.(y_val)
-
+function evaluate_epoch(model, x_train, forcings_train, y_train, mask_train, x_val, forcings_val, y_val, mask_val, ps, st, init::EpochSnapshot, cfg::TrainConfig)
+    ps_cpu = ps |> cfg.cdev
```
[the remark this answers is not in the document]
Comment (lazarusA) [reply src/training/epoch.jl:54]: shouldn't evaluation still happen on gpu? if available... I think so.

File: src/training/epoch.jl [hunk 12]
```diff
@@ -1,28 +1,35 @@
-function run_epoch!(loader, model, ps, st, opt_state, cfg::TrainConfig)
-    loss_fn = build_loss_fn(model, cfg)
+function collect_dim_data(x, y, cfg)
+    x_col = Array(x[1])
+    forcing_nt = NamedTuple([k => Array(v) for (k, v) in pairs(x[2])])
+    targets_nt = NamedTuple([k => Array(v) for (k, v) in pairs(y[1])])
+    masks_nt = NamedTuple([k => Array(v) for (k, v) in pairs(y[2])])
```
[the remark this answers is not in the document]
Comment (Qfl3x) [reply src/training/epoch.jl]: The reason behind is is if the inputs are DimArrays they need to be collected explicitly, thus the `Array` calls

**Comment by lazarusA** [comment]
the generic hybrid for multi neural networks is still failing 
https://earthyscience.github.io/EasyHybrid.jl/dev/research/synthetic_respiration

as well as split sequences for lstms (https://earthyscience.github.io/EasyHybrid.jl/dev/tutorials/example_synthetic_lstm)

First MultiNNs

```Julia
using EasyHybrid
df = load_timeseries_netcdf("https://github.com/bask0/q10hybrid/raw/master/data/Synthetic4BookChap.nc");
dfnot = Float32.(df[!, Not(:time)]);

ka = to_keyedArray(dfnot);

using DimensionalData
mat = Array(Matrix(dfnot)')
da = DimArray(mat, (Dim{:variable}(Symbol.(names(dfnot))), Dim{:batch_size}(1:size(dfnot, 1))));

function RbQ10(; ta, Q10, rb, tref = 15.0f0)
    reco = rb .* Q10 .^ (0.1f0 .* (ta .- tref))
    return (; reco, Q10, rb)
end

parameters = (
    # Parameter name | Default | Lower | Upper      | Description
    rb = (3.0f0, 0.0f0, 13.0f0),  # Basal respiration [μmol/m²/s]
    Q10 = (2.0f0, 1.0f0, 4.0f0),   # Temperature sensitivity factor [-]
)

forcing = [:ta]                    # Forcing variables (temperature)
target = [:reco]                   # Target variable (respiration)

global_param_names = [:Q10]        # Global parameters (same for all samples)
neural_param_names = [:rb]         # Neural network predicted parameters


predictors_multi_nn = (rb = [:sw_pot, :dsw_pot],)

multi_nn_hybrid_model = constructHybridModel(
    predictors_multi_nn,              # Input features
    forcing,                 # Forcing variables
    target,                  # Target variables
    RbQ10,                  # Process-based model function
    parameters,              # Parameter definitions
    global_param_names,      # Global parameters
    hidden_layers = [16, 16], # Neural network architecture
    activation = sigmoid,      # Activation function
    scale_nn_outputs = true, # Scale neural network outputs
    input_batchnorm = true   # Apply batch normalization to inputs
)

multi_nn_out = train(
    multi_nn_hybrid_model,
    ka,
    ();
    nepochs = 10,           # Number of training epochs
    batchsize = 512,         # Batch size for training
    opt = AdamW(0.1),   # Optimizer and learning rate
    monitor_names = [:rb, :Q10], # Parameters to monitor during training
    yscale = identity,       # Scaling for outputs
    show_progress = false,
    shuffleobs = true,
    model_name = "RbQ10_synthetic4"
)

LuxCore.testmode(multi_nn_out.st)
mean(df.dsw_pot)
mean(df.sw_pot)
```

File: src/io/save.jl [hunk 13]
```diff
@@ -53,7 +53,7 @@ function save_observations!(file_name, target_names, yobs, train_or_val_name)
 end
 
 function to_named_tuple(ka, target_names)
-    arrays = [Array(ka(variable = k)) for k in target_names]
+    arrays = [Array(ka[k]) for k in target_names]
```
[the remark this answers is not in the document]
Comment (BernhardAhrens) [reply src/io/save.jl:56]: @Qfl3x Not sure if I have an example or test for LSTMs but I believe `[Array(ka(variable = k)) for k in target_names]` this was required

**Comment by BernhardAhrens** [comment]
@Qfl3x Yes, the building the documentation should basically be also seen as a test

**Comment by BernhardAhrens** [comment]
Now only the LSTM is failing. The thing in the other tutorials was partly plotting

**Comment by lazarusA** [comment]
continues in #259
